### PR TITLE
- Add reportDIETER

### PR DESCRIPTION
**Pull request by pratik98**

### CODE REVIEW

File: R/reportDIETER.R [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+    
+    out_h <- rep_hrs %>% 
+      dplyr::select_(~model, ~year, ~variable, ~country,~hour,~value) %>%
+      dplyr::mutate_(hour = ~as.numeric(stringr::str_extract(hour, "[0-9]+"))) %>% 
+      dplyr::group_by_(~model, ~variable, ~year, ~country) %>%
+      complete(hour = (1:8760)) %>%
+      #replace(is.na(.), 0) %>%
+      dplyr::mutate_all(coalesce, 0)%>%
+      dplyr::ungroup(c('model', 'variable', 'year', 'country')) %>%
+      dplyr::mutate_(Model = ~model, Scenario = ~paste0("baseline"), Region = ~country,
```
Comment (pfuehrlich-pik) [inline R/reportDIETER.R:49]: This paste is not doing anything, is it?
```suggestion
      dplyr::mutate_(Model = ~model, Scenario = ~baseline, Region = ~country,
```

File: R/reportDIETER.R [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+      dplyr::select_(~Model,~Scenario,~Region,~Variable,~Year,~Period,~Tech,~Value,~Hour)
+    
+    ###################################################################
+    rep_techHrs <- read.gdx(gdxName = file, requestList.name = 'report_tech_hours', factors = FALSE, squeeze = FALSE) 
+    
+    names(rep_techHrs) <- c("gdxfile", "model","year","country","variable", "tech", "hour", "value")
+    
+    out_th <- rep_techHrs %>% 
+      dplyr::select_(~model, ~year, ~tech,~variable, ~country, ~hour,~value) %>%
+      dplyr::mutate_(hour = ~as.numeric(stringr::str_extract(hour, "[0-9]+"))) %>%
```
Comment (pfuehrlich-pik) [inline R/reportDIETER.R:64]: Replace `stringr::str_extract` with sub as discussed ealier.

File: R/reportDIETER.R [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+      dplyr::mutate_(hour = ~as.numeric(stringr::str_extract(hour, "[0-9]+"))) %>% 
+      dplyr::mutate_(tech = ~as.character(tech)) %>%
+      dplyr::group_by_(~model, ~tech, ~hour, ~variable, ~country) %>%
+      dplyr::mutate_(year = ~as.numeric(year)) %>% 
+      complete(year = c(2010,2015,2020,2025,2030,2035,2040,2045,2050,2055,2060,2070,2080,2090,2100)) %>%
+      
+      dplyr::ungroup(c('model', 'tech', 'hour', 'variable', 'country')) %>% 
+      dplyr::group_by_(~model, ~year, ~variable, ~country,~tech) %>%
+      complete(hour = (1:8760)) %>%
+      #replace(is.na(.), 0) %>%
```
Comment (pfuehrlich-pik) [inline R/reportDIETER.R:73]: Remove if no longer needed
```suggestion
```

File: R/reportDIETER.R [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+    annual_new$`2130`<- NA
+  }
+  if(!'2150' %in% colnames(annual_new)){
+    annual_new$`2150` <- NA
+  }
+  
+  
+  #technology mapping
+  dieter.tech.mapping <- c(CCGT = "CCGT",
+                           #lig = "Lignite",
```
Comment (pfuehrlich-pik) [inline R/reportDIETER.R:179]: Please remove commented code, and maybe add a comment explaining that some things are still missing here or will come in the future.
```suggestion
```

File: R/reportDIETER.R [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+                           #coal = "Coal (Lig + HC)",
+  )
+  
+  
+  
+  dieter.tech.mapping.standalone <- c(dieter.tech.mapping,
+                                      lig = "Lignite",
+                                      hc = "Hard coal",
+                                      coal = "Coal (Lig + HC)",
+                                      NULL)
```
Comment (pfuehrlich-pik) [inline R/reportDIETER.R:196]: This NULL is ignored by `c()`
```suggestion
                                      )
```

File: R/reportDIETER.R [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+  
+  annual_new$Tech <- as.factor(annual_new$Tech)
+  annual_remind <-  annual_new
+  annual_new <- annual_new %>% quitte::revalue.levels(Tech = dieter.tech.mapping.standalone)
+  
+  
+  annual_new$Variable <- paste0(stringr::str_trim(annual_new$Variable,side = c("both")),"|",annual_new$Tech)
```
Comment (pfuehrlich-pik) [inline R/reportDIETER.R:203]: Please test this, I just checked `gsub('^\\W+|\\W+$', '', ' \t asd \na\n \n') == 'asd \na'`
```suggestion
  annual_new$Variable <- paste0(gsub("^\\W+|\\W+$", "", annual_new$Variable), "|", annual_new$Tech)
```

File: R/reportDIETER.R [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+  read.quitte(file.path(outputDir, subFolder,"Dieter_Annual.mif")) %>% 
+    readr::write_rds(paste0(gsub(".mif","",file.path(outputDir, subFolder,"Dieter_Annual.mif")),'.rds'), compress = 'xz') 
+  
+  
+  # append model = 'DIETER' to REMIND-EU mif file
+  # then for the switch that appends to the main mif, use coal = "Coal (Lig + HC)", 
+  
+  dieter.tech.mapping.REMIND <- c(dieter.tech.mapping,
+                                  coal = "Coal (Lig + HC)",
+                                  NULL)
```
Comment (pfuehrlich-pik) [inline R/reportDIETER.R:222]: NULL is ignored by `c()`
```suggestion
                                  )
```

File: R/reportDIETER.R [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+  data.table::setDT(annual_remind)
+  
+  # append to Main REMIND file
+  # load main mif files
+  remind_miffiles <- list.files(outputDir, pattern = "REMIND_generic", full.names = F)
+  #miffile_name <- name_mif[!grepl("withoutPlu", name_mif)]
+  for (mif_file in remind_miffiles) {
+    
+    name_mif <- file.path(outputDir, mif_file)
+    stopifnot(typeof(name_mif) == "character")
```
Comment (pfuehrlich-pik) [inline R/reportDIETER.R:242]:
```suggestion
    stopifnot(is.character(name_mif))
```

File: R/reportDIETER.R [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+#' @author Chen Gong Pratik Agrawal
+#'
+
+#' @importFrom data.table fread fwrite setDT
+#' @importFrom quitte read.quitte read.gdx revalue.levels
+#' @importFrom readr write_rds
+#' @importFrom rlang .data
+#' @importFrom dplyr %>% left_join filter select rename mutate arrange ungroup across mutate_all coalesce
+#' @importFrom tidyr complete
+#' @importFrom stringr str_extract str_trim
```
Comment (pfuehrlich-pik) [inline R/reportDIETER.R:19]:
```suggestion
```

File: R/reportDIETER.R [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+  
+  dieter.tech.mapping.REMIND <- c(dieter.tech.mapping,
+                                  coal = "Coal (Lig + HC)",
+                                  NULL)
+  
+  annual_remind <- annual_remind %>%dplyr::filter_(~Model == "DIETER",~Tech !='lig',~Tech !='hc')
+  annual_remind <- annual_remind %>% quitte::revalue.levels(Tech = dieter.tech.mapping.REMIND)
+  
+
+  annual_remind$Variable <- paste0(stringr::str_trim(annual_remind$Variable,side = c("both")),"|",annual_remind$Tech)
```
Comment (pfuehrlich-pik) [inline R/reportDIETER.R:228]: Replace stringr with gsub, as discussed

File: R/reportDIETER.R [hunk 11]
```diff
@@ -0,0 +1,284 @@
+#' Reporting for the coupled DIETER Model
+#'
+#'
+#' *Warning* The function modifies the "REMIND_generic_<scenario>.mif" file by appending the
```
[the remark this answers is not in the document]
Comment (pfuehrlich-pik) [reply R/reportDIETER.R:4]: @LaviniaBaumstark What's the best way to do this, or is it fine as is? The actual appending happens in line 246 I believe.

File: R/reportDIETER.R [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+#' additional reporting variables.
+#'
+#' 
+#' @param dieterDatafile full path with name of dieter gdx file.
+#' @param outputDir path to the output folder, default is current folder.
+#' @author Chen Gong Pratik Agrawal
```
Comment (pfuehrlich-pik) [inline R/reportDIETER.R:10]:
```suggestion
#' @author Chen Gong, Pratik Agrawal
```

File: R/reportDIETER.R (hunk 11 above)
[the remark this answers is not in the document]
Comment (LaviniaBaumstark) [reply R/reportDIETER.R:4]: why do you need the DIETER reporting in the REMIND_generic_*.mif? i would prefer a separate reporting file

[the remark this answers is not in the document]
Comment (pratik98) [reply R/reportDIETER.R:4]: The DIETER data is appended to a REMIND_generic_*.mif file  and  the new mif is saved in different location, i.e sub directory "DIETER". The original REMIND_generic_*.mif  file is not modified.

**Pull request merged**